src/Functions/seriesDecomposeSTL.cpp

```diff
@@ -0,0 +1,264 @@
+
+    std::string getName() const override { return name; }
+
+    size_t getNumberOfArguments() const override { return 2; }
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        FunctionArgumentDescriptors args{
+            {"time-series", &isArray<IDataType>, nullptr, "Array"},
+            {"period", &isNativeNumber<IDataType>, nullptr, "Number"},
+        };
+        validateFunctionArgumentTypes(*this, arguments, args);
+
+        return std::make_shared<DataTypeArray>(std::make_shared<DataTypeArray>(std::make_shared<DataTypeFloat32>()));
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t) const override
+    {
+        ColumnPtr array_ptr = arguments[0].column;
+        const ColumnArray * array = checkAndGetColumn<ColumnArray>(array_ptr.get());
+        if (!array)
+        {
+            const ColumnConst * const_array = checkAndGetColumnConst<ColumnArray>(
+                arguments[0].column.get());
+            if (!const_array)
+                throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Illegal column {} of first argument of function {}",
+                    arguments[0].column->getName(), getName());
+
+            array_ptr = const_array->convertToFullColumn();
+            array = assert_cast<const ColumnArray *>(array_ptr.get());
+        }
+
+        const IColumn & src_data = array->getData();
+        const ColumnArray::Offsets & src_offsets = array->getOffsets();
+
+        Float64 period;
+
+        auto ret = ColumnFloat32::create();
+        auto & res_data = ret->getData();
+
+        ColumnArray::ColumnOffsets::MutablePtr col_offsets = ColumnArray::ColumnOffsets::create();
+        auto & col_offsets_data = col_offsets->getData();
+
+        auto root_offsets = ColumnArray::ColumnOffsets::create();
+        auto & root_offsets_data = root_offsets->getData();
+
+        ColumnArray::Offset prev_src_offset = 0;
+
+        for (size_t i = 0; i < src_offsets.size(); ++i)
+        {
+            auto period_ptr = arguments[1].column->convertToFullColumnIfConst();
+            if (checkAndGetColumn<ColumnUInt8>(period_ptr.get()) || checkAndGetColumn<ColumnUInt16>(period_ptr.get())
+                || checkAndGetColumn<ColumnUInt32>(period_ptr.get()) || checkAndGetColumn<ColumnUInt64>(period_ptr.get()))
+                period = period_ptr->getUInt(i);
+            else if (checkAndGetColumn<ColumnFloat32>(period_ptr.get()) || checkAndGetColumn<ColumnFloat64>(period_ptr.get()))
+            {
+                period = period_ptr->getFloat64(i);
+                if (isNaN(period) || !std::isfinite(period) || period < 0)
```

Comment:
I am pretty sure we can use `std::isnan()` directly instead of `isNaN()` (period is already a FP-value and no fallback handling for non-FP types is needed).